Lesson3

artursmolin opened the pull request:
1. Создайте Rails-модели с нужными атрибутами для следующих сущностей:
Тест
Вопрос
Ответ
Пользователь
Категория
2. С помощью механизма миграций, добавьте ограничения целостности на уровне БД в соответствующих таблицах для необходимых атрибутов так, чтобы они не могли содержать значение NULL
3. С помощью механизма миграций добавьте значения по умолчанию:
Для уровня Теста
Для атрибута correct у Ответа

artursmolin committed:
add models

artursmolin committed:
add migrations

artursmolin committed:
fix

max-underthesun commented on test_guru/db/migrate/20190401110324_create_tests.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :tests do |t|
+      t.string :title
+      t.integer :level
+      t.integer :category_id
```
в задании по SQL нужно было создавать не просто поля, а [внешние ключи](https://en.wikipedia.org/wiki/Foreign_key)
база данных в этом случае применяет к этому полю `FOREIGN KEY constraint` для контроля целостности данных ([Data integrity](https://en.wikipedia.org/wiki/Data_integrity))

рекомендуется всегда обозначать для полей-связей, что это ссылка на внешний ключ
так же рекомендуется создавать на них индексы для ускорения выборки
у тебя там ошибка в задании по SQL
аналогичная ошибка здесь - это создано просто поле, но не внешний ключ
попробуй разобраться

max-underthesun replied:
аналогичное замечание по всем ссылкам на другие таблицы во всех миграциях

max-underthesun commented on test_guru/db/migrate/20190401112529_add_tests_level_null_constrait.rb:
```diff
@@ -0,0 +1,12 @@
+
+  def down
+    change_column_null(:tests, :level, true)
+  end
```
насколько я помню,  `change_column_null` - обратимая миграция
поэтому здесь достаточно в таком духе написать:
```ruby
  def change
    change_column_null(:tests, :level, false)
  end
```

max-underthesun replied:
аналогично в остальных миграциях

max-underthesun replied:
и удобней сгруппировать миграции
к примеру, если меняется несколько колонок в одной таблице - запиши все в одной мгирации

max-underthesun commented on test_guru/db/migrate/20190401114823_add_answers_correct_default_value.rb:
```diff
@@ -0,0 +1,12 @@
+  def down
+    change_column :answers, :correct, :boolean, default: nil
+  end
+end
```
в [документации](https://edgeguides.rubyonrails.org/active_record_migrations.html#changing-columns) рекомендуют записывать миграцию смены "дефолта" вот так:
```ruby
change_column_default :products, :approved, from: true, to: false
```

max-underthesun commented on test_guru/db/migrate/20190401110337_create_answers.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :answers do |t|
+      t.string :body
+      t.integer :question_id
+      t.boolean :correct
```
для логических полей рекомендуется указывать опцию `null: false`
тогда в базе будут возможны строго два значения `true` или `false` вместо трех (`null/true/false`)

artursmolin committed:
ref some

artursmolin committed:
fix migrations

fix foreign key problem

artursmolin committed:
fix column_default

max-underthesun commented on test_guru/db/migrate/20190401112529_add_tests_level_null_constrait.rb:
```diff
@@ -0,0 +1,14 @@
+  end
+
+  def down
+    change_column_default :tests, :level, from:'Junior', to: nil
```
посмотри еще раз документацию
`change_column_default` с использованием `from:`, `to:` обратимая и не нужно писать явно `up` и `down`

max-underthesun commented on test_guru/db/migrate/20190401113002_add_questions_test_id_null_constrait.rb:
```diff
@@ -0,0 +1,5 @@
+class AddQuestionsTestIdNullConstrait < ActiveRecord::Migration[5.2]
+  def change
+    add_reference :questions, :tests, foreign_key: true
```
обрати внимание на синтаксис [метода](https://api.rubyonrails.org/v5.2/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-references)
вторым аргументом должно идти не имя таблицы (множественное число), а имя связи (единственное число)

max-underthesun replied:
аналогичная ошибка в других миграциях... и посмотри в схеме, у тебя вместо ключа `test_id` создался ключ `tests_id`, что неверно

max-underthesun commented on test_guru/db/migrate/20190401113030_add_answers_question_id_null_constrait.rb:
```diff
@@ -0,0 +1,13 @@
+  end
+
+  def down
+    change_column_default :answers, :correct, from: true ,to:nil 
```
опять ненужные "ап" и "даун"

max-underthesun reviewed (commented):
`NULL` можно (и лучше сделать иманно так) запретить для большинства полей, а у тебя сейчас вижу только два поля имеют этот запрет

artursmolin committed:
fix

1. null добавлен ко всем полям, кроме e-mail
2. убрали методы up,down
3. поправил внешние ключи

max-underthesun commented on test_guru/db/migrate/20190401110343_create_users.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :users do |t|
+      t.string :name, null: false
+      t.string :email
```
в большинстве случаев именно `email` является обязательным полем сущности Пользователь, как раз этому аттрибуту стоило добавить "нул констрейн"

max-underthesun commented on test_guru/db/migrate/20190401112529_add_tests_level_null_constrait.rb:
```diff
@@ -0,0 +1,7 @@
+class AddTestsLevelNullConstrait < ActiveRecord::Migration[5.2]
+  def change
+    change_column_null(:tests, :level, false)
```
раз уж все остальные без скобок, писал бы и здесь без скобок )
